core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -0,0 +1,397 @@
+  * limitations under the License.
+  */
+
+package kafka.api
+
+import java.util.Properties
+
+import kafka.integration.KafkaServerTestHarness
+import kafka.server.KafkaConfig
+import kafka.utils.TestUtils
+import org.apache.kafka.clients.consumer.{ConsumerConfig, ConsumerRecord, KafkaConsumer}
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.KafkaException
+import org.apache.kafka.common.errors.ProducerFencedException
+import org.apache.kafka.common.protocol.SecurityProtocol
+import org.junit.{After, Before, Ignore, Test}
+import org.junit.Assert._
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.ExecutionException
+import scala.util.Random
+
+class TransactionsTest extends KafkaServerTestHarness {
+  val numServers = 3
+  val topic1 = "topic1"
+  val topic2 = "topic2"
+
+  override def generateConfigs : Seq[KafkaConfig] = {
+    TestUtils.createBrokerConfigs(numServers, zkConnect, true).map(KafkaConfig.fromProps(_, serverProps()))
+  }
+
+  @Before
+  override def setUp(): Unit = {
+    super.setUp()
+    val numPartitions = 3
+    val topicConfig = new Properties();
+    topicConfig.put(KafkaConfig.MinInSyncReplicasProp, 2.toString)
+    TestUtils.createTopic(zkUtils, topic1, numPartitions, numServers, servers, topicConfig)
+    TestUtils.createTopic(zkUtils, topic2, numPartitions, numServers, servers, topicConfig)
+  }
+
+  @After
+  override def tearDown(): Unit = {
+    super.tearDown()
+  }
+
+  @Test
+  def testBasicTransactions() = {
+    val producer = TestUtils.createTransactionalProducer("my-hello-world-transactional-id", servers)
+    val consumer = transactionalConsumer("transactional-group")
+    val unCommittedConsumer = nonTransactionalConsumer("non-transactional-group")
+    try {
+      producer.initTransactions()
+
+      producer.beginTransaction()
+      producer.send(producerRecord(topic2, "2", "2", willBeCommitted = false))
+      producer.send(producerRecord(topic1, "4", "4", willBeCommitted = false))
+      producer.abortTransaction()
+
+      producer.beginTransaction()
+      producer.send(producerRecord(topic1, "1", "1", willBeCommitted = true))
+      producer.send(producerRecord(topic2, "3", "3", willBeCommitted = true))
+      producer.commitTransaction()
+
+      consumer.subscribe(List(topic1, topic2))
+      unCommittedConsumer.subscribe(List(topic1, topic2))
+
+      val records = pollUntilExactlyNumRecords(consumer, 2)
+      records.zipWithIndex.foreach { case (record, i) =>
+        assertCommittedAndGetValue(record)
+      }
+
+      val allRecords = pollUntilExactlyNumRecords(unCommittedConsumer, 4)
+      val expectedValues = List("1", "2", "3", "4").toSet
+      allRecords.zipWithIndex.foreach { case (record, i) =>
+        assertTrue(expectedValues.contains(recordValue(record)))
+      }
+    } catch {
+      case e @ (_ : KafkaException | _ : ProducerFencedException) =>
```

Comment:
nit: do we need this? An unexpected exception will cause the test to fail anyway. Same for the other tests.